tasks/process_flakes.py

```diff
@@ -0,0 +1,107 @@
+import datetime as dt
+import logging
+from typing import Any
+
+from shared.celery_config import process_flakes_task_name
+from shared.django_apps.reports.models import Flake, TestInstance
+
+from app import celery_app
+from tasks.base import BaseCodecovTask
+
+log = logging.getLogger(__name__)
+
+
+FlakeDict = dict[Any, Flake]
+
+FLAKE_EXPIRY_COUNT = 30
+
+
+class ProcessFlakesTask(BaseCodecovTask, name=process_flakes_task_name):
+    """
+    This task is currently called in the test results finisher task and in the sync pulls task
+    """
+
+    def run_impl(
+        self,
+        _db_session,
+        *,
+        repo_id,
+        commit_id_list,
+        branch,
+        **kwargs,
+    ):
+        repo_id = int(repo_id)
+        log.info(
+            "Received process flakes task",
+            extra=dict(repoid=repo_id, commit=commit_id_list),
+        )
+
+        flake_dict = generate_flake_dict(repo_id)
+
+        for commit_id in commit_id_list:
+            test_instances = get_test_instances(commit_id, repo_id, branch)
+            for test_instance in test_instances:
+                if test_instance.outcome == TestInstance.Outcome.PASS.value:
+                    flake = flake_dict.get(test_instance.test_id)
+                    if flake is not None:
+                        update_passed_flakes(flake)
+                elif (
+                    test_instance.outcome == TestInstance.Outcome.FAILURE.value
+                    or test_instance.outcome == TestInstance.Outcome.ERROR.value
+                ):
+                    flake = flake_dict.get(test_instance.test_id)
+                    upserted_flake = upsert_failed_flake(test_instance, repo_id, flake)
+                    if flake is None:
+                        flake_dict[upserted_flake.test_id] = upserted_flake
+
+        return {"successful": True}
+
+
+def get_test_instances(commit_id: str, repo_id: int, branch: str) -> list[TestInstance]:
+    test_instances = list(
+        TestInstance.objects.filter(
+            commitid=commit_id, repoid=repo_id, branch=branch
+        ).all()
+    )
+    return test_instances
+
+
+def generate_flake_dict(repo_id: int) -> FlakeDict:
+    flakes = Flake.objects.filter(repository_id=repo_id, end_date__isnull=True).all()
```

Comment:
this query may need a separate index, because while we may be using the `reduced_error_id` field in the future we will definitely not be using the `test_id`